File: backend/src/main/java/com/smartcampus/model/Resource.java
package com.smartcampus.model;

import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.relational.core.mapping.Table;

import java.time.Instant;
import java.util.UUID;

@Table("resources")
public record Resource(
        @Id UUID id,
        UUID domainId,
        String resourceType, // stored as VARCHAR — use ResourceType.valueOf() to convert
        String name,
        String description,
        String location,
        Integer capacity,
        String status, // stored as VARCHAR — use ResourceStatus.valueOf() to convert
        String metadata,
        UUID createdBy,
        @CreatedDate Instant createdAt,
        @LastModifiedDate Instant updatedAt) {
    /**
     * Convenience accessor — never throws if DB value is always written via
     * ResourceStatus.name()
     */
    public ResourceStatus statusEnum() {
        return ResourceStatus.valueOf(this.status);
    }

    public ResourceType resourceTypeEnum() {
        return ResourceType.valueOf(this.resourceType);
    }
}

File: frontend/src/types/resource.ts

export type ResourceType =
    | 'LECTURE_HALL'
    | 'LAB'
    | 'MEETING_ROOM'
    | 'EQUIPMENT'
    | 'OTHER';

export type ResourceStatus = 'ACTIVE' | 'OUT_OF_SERVICE';

export interface ResourceResponse {
    id: string;
    domainId: string;
    domainName: string | null;
    resourceType: ResourceType;
    name: string;
    description: string | null;
    location: string;
    capacity: number | null;
    status: ResourceStatus;
    metadata: string | null;
    createdBy: string;
    createdAt: string;
    updatedAt: string;
}

export interface CreateResourceRequest {
    domainId: string;
    resourceType: ResourceType;
    name: string;
    description?: string;
    location: string;
    capacity?: number;
    metadata?: string;
}

export interface UpdateResourceRequest {
    resourceType: ResourceType;
    name: string;
    description?: string;
    location: string;
    capacity?: number;
    metadata?: string;
}

export interface UpdateResourceStatusRequest {
    status: ResourceStatus;
}

export interface ResourceSearchParams {
    domainId?: string;
    resourceType?: string;
    status?: string;
    minCapacity?: number;
    location?: string;
    query?: string;
}